File: crates/kuben-api/src/routes/vulnerabilities.rs
//! Vulnerability exceptions of an organization (M4.6): one finding may pass
//! the scan gates until the exception expires or is revoked. Granting one
//! weakens every gate it touches, so only people with organization admin
//! rights do it, name an owner and a reason, and never for longer than 90
//! days. Tokens cannot grant or revoke.

use axum::{
    Json,
    extract::{Path, Query, State},
    http::StatusCode,
};
use kuben_core::{Error, authz::ScopeChain, ids::OrgId, perm::Perm, scan::MAX_EXCEPTION_SECS, time::now_ms};
use kuben_store::repo::{NewException, VulnException};
use serde::{Deserialize, Serialize};
use utoipa::{IntoParams, ToSchema};
use uuid::Uuid;

use super::{request, scope};
use crate::{
    authz::Authz,
    error::{ApiError, ApiResult},
    state::ApiState,
};

#[derive(Debug, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct CreateException {
    /// The finding's id, e.g. `CVE-2026-12345` or `GHSA-xxxx-xxxx-xxxx`.
    #[schema(example = "CVE-2026-12345")]
    pub vulnerability: String,
    pub reason: String,
    /// Who answers for it: a person or a team.
    #[schema(example = "platform-team@example.com")]
    pub owner: String,
    /// 1 to 90 days.
    #[schema(example = 30)]
    pub days: u32,
    /// Limit it to one project; every project when omitted.
    pub project: Option<String>,
}

#[derive(Debug, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct ExceptionDto {
    pub id: Uuid,
    pub vulnerability: String,
    pub project: Option<Uuid>,
    pub reason: String,
    pub owner: String,
    pub created_by: String,
    pub created_at: String,
    pub expires_at: String,
    pub revoked_at: Option<String>,
    /// In force now.
    pub active: bool,
}

#[derive(Debug, Deserialize, IntoParams)]
#[into_params(parameter_in = Query)]
pub struct ListQuery {
    /// Include expired and revoked exceptions.
    #[serde(default)]
    pub all: bool,
}

impl ExceptionDto {
    fn of(e: VulnException, now: i64) -> Self {
        Self {
            active: e.revoked_at.is_none() && e.expires_at > now,
            id: e.id,
            vulnerability: e.vulnerability,
            project: e.project.map(|p| *p.as_uuid()),
            reason: e.reason,
            owner: e.owner,
            created_by: e.created_by,
            created_at: request::timestamp(e.created_at),
            expires_at: request::timestamp(e.expires_at),
            revoked_at: e.revoked_at.map(request::timestamp),
        }
    }
}

fn org_of(authz: &Authz) -> Result<OrgId, ApiError> {
    authz.org_ids().first().copied().ok_or(ApiError(Error::Forbidden))
}

fn check(body: &CreateException) -> Result<(), Error> {
    let id = body.vulnerability.trim();
    let valid_id = !id.is_empty()
        && id.len() <= 64
        && id.as_bytes()[0].is_ascii_alphanumeric()
        && id
            .bytes()
            .all(|b| b.is_ascii_alphanumeric() || matches!(b, b'.' | b'_' | b':' | b'-'));
    if !valid_id {
        return Err(Error::Validation(format!("`{id}` is not a vulnerability id")));
    }
    let text = |name: &str, value: &str, max: usize| {
        let value = value.trim();
        if value.is_empty() || value.chars().count() > max || value.chars().any(char::is_control) {
            Err(Error::Validation(format!(
                "{name} must be 1 to {max} printable characters"
            )))
        } else {
            Ok(())
        }
    };
    text("reason", &body.reason, 1024)?;
    text("owner", &body.owner, 256)?;
    let max_days = MAX_EXCEPTION_SECS / 86_400;
    if body.days == 0 || i64::from(body.days) > max_days {
        return Err(Error::Validation(format!(
            "an exception lasts 1 to {max_days} days"
        )));
    }
    Ok(())
}

/// The organization's vulnerability exceptions in force (or all).
#[utoipa::path(
    get,
    path = "/vulnerability-exceptions", operation_id = "listVulnerabilityExceptions",
    tag = "security",
    params(ListQuery),
    responses((status = 200, body = Vec<ExceptionDto>), (status = 403, body = crate::error::Problem))
)]
pub async fn list(
    State(state): State<ApiState>,
    authz: Authz,
    Query(q): Query<ListQuery>,
) -> ApiResult<Json<Vec<ExceptionDto>>> {
    let org = org_of(&authz)?;
    let _proof = authz.require(&state, Perm::OrgRead, &ScopeChain::org(org))?;
    let mut tenant = state.store.tenant(org).await?;
    let now = now_ms();
    let found = tenant.exceptions(q.all).await?;
    Ok(Json(
        found.into_iter().map(|e| ExceptionDto::of(e, now)).collect(),
    ))
}

/// Let one finding pass the scan gates for a while.
#[utoipa::path(
    post,
    path = "/vulnerability-exceptions", operation_id = "createVulnerabilityException",
    tag = "security",
    request_body = CreateException,
    responses(
        (status = 201, body = ExceptionDto),
        (status = 403, body = crate::error::Problem),
        (status = 404, body = crate::error::Problem),
        (status = 422, body = crate::error::Problem),
    )
)]
pub async fn create(
    State(state): State<ApiState>,
    authz: Authz,
    Json(body): Json<CreateException>,
) -> ApiResult<(StatusCode, Json<ExceptionDto>)> {
    authz.forbid_token()?;
    let org = org_of(&authz)?;
    let _proof = authz.require(&state, Perm::OrgAdmin, &ScopeChain::org(org))?;
    check(&body)?;
    let project = match body.project.as_deref() {
        None => None,
        Some(slug) => Some(scope::project(&state, &authz, slug).await?.id()),
    };
    let (_, actor) = request::actor(&authz);
    let now = now_ms();
    let new = NewException {
        project,
        vulnerability: body.vulnerability.trim().to_owned(),
        reason: body.reason.trim().to_owned(),
        owner: body.owner.trim().to_owned(),
        created_by: actor,
        expires_at: now + i64::from(body.days) * 86_400_000,
    };
    let mut tenant = state.store.tenant(org).await?;
    let id = tenant.create_exception(&new).await?;
    let mut audit = request::audit(
        &authz,
        "vulnerability.exception.created",
        "vulnerability",
        new.vulnerability.clone(),
    );
    audit.data = Some(serde_json::json!({ "id": id, "owner": new.owner, "expires_at": new.expires_at }));
    tenant.append_audit(audit).await?;
    let made = tenant
        .exceptions(true)
        .await?
        .into_iter()
        .find(|e| e.id == id)
        .ok_or_else(|| Error::Internal("the new exception is missing".into()))?;
    tenant.commit().await?;
    Ok((StatusCode::CREATED, Json(ExceptionDto::of(made, now))))
}

/// Revoke an exception for good.
#[utoipa::path(
    delete,
    path = "/vulnerability-exceptions/{id}", operation_id = "revokeVulnerabilityException",
    tag = "security",
    params(("id" = Uuid, Path, description = "Exception id")),
    responses(
        (status = 204, description = "Revoked"),
        (status = 403, body = crate::error::Problem),
        (status = 404, body = crate::error::Problem),
    )
)]
pub async fn revoke(
    State(state): State<ApiState>,
    authz: Authz,
    Path(id): Path<Uuid>,
) -> ApiResult<StatusCode> {
    authz.forbid_token()?;
    let org = org_of(&authz)?;
    let _proof = authz.require(&state, Perm::OrgAdmin, &ScopeChain::org(org))?;
    let (_, actor) = request::actor(&authz);
    let mut tenant = state.store.tenant(org).await?;
    if !tenant.revoke_exception(id, &actor).await? {
        return Err(Error::NotFound(format!("exception `{id}`")).into());
    }
    let audit = request::audit(
        &authz,
        "vulnerability.exception.revoked",
        "vulnerability",
        id.to_string(),
    );
    tenant.append_audit(audit).await?;
    tenant.commit().await?;
    Ok(StatusCode::NO_CONTENT)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn body(vulnerability: &str, days: u32) -> CreateException {
        CreateException {
            vulnerability: vulnerability.into(),
            reason: "not reachable from the network".into(),
            owner: "platform@example.com".into(),
            days,
            project: None,
        }
    }

    #[test]
    fn exceptions_are_bounded() {
        assert!(check(&body("CVE-2026-12345", 30)).is_ok());
        assert!(check(&body("GHSA-abcd-1234-efgh", 90)).is_ok());
        for bad in [
            body("", 1),
            body("-CVE", 1),
            body("CVE 1", 1),
            body("CVE-1", 0),
            body("CVE-1", 91),
        ] {
            assert!(check(&bad).is_err(), "{bad:?}");
        }
        let ownerless = CreateException {
            owner: "  ".into(),
            ..body("CVE-1", 1)
        };
        assert!(check(&ownerless).is_err());
    }
}

File: internal/httpapi/vulnerabilities.go
package api

// Vulnerability exceptions of an organization (M4.6): one finding may pass
// the scan gates until the exception expires or is revoked. Granting one
// weakens every gate it touches, so only people with organization admin
// rights do it, name an owner and a reason, and never for longer than 90
// days. Tokens cannot grant or revoke (routes/vulnerabilities.rs).

import (
	"context"
	"strings"
	"unicode"
	"unicode/utf8"

	"github.com/Teamtem-dev/kuben/internal/core/authz"
	"github.com/Teamtem-dev/kuben/internal/core/clock"
	"github.com/Teamtem-dev/kuben/internal/core/ids"
	"github.com/Teamtem-dev/kuben/internal/core/kerrors"
	"github.com/Teamtem-dev/kuben/internal/core/opt"
	"github.com/Teamtem-dev/kuben/internal/core/perm"
	"github.com/Teamtem-dev/kuben/internal/core/scan"
	"github.com/Teamtem-dev/kuben/internal/httpapi/access"
	"github.com/Teamtem-dev/kuben/internal/httpapi/gen"
	"github.com/Teamtem-dev/kuben/internal/store"
)

// exceptionDto is ExceptionDto::of at now.
func exceptionDto(e store.VulnException, now int64) gen.ExceptionDto {
	project := nullUUID()
	if p, ok := e.Project.Get(); ok {
		project = gen.NewOptNilUUID(p.UUID())
	}
	revoked := opt.None[string]()
	if at, ok := e.RevokedAt.Get(); ok {
		revoked = opt.Some(Timestamp(at))
	}
	return gen.ExceptionDto{
		Active:        e.RevokedAt.IsNone() && e.ExpiresAt > now,
		ID:            e.ID,
		Vulnerability: e.Vulnerability,
		Project:       project,
		Reason:        e.Reason,
		Owner:         e.Owner,
		CreatedBy:     e.CreatedBy,
		CreatedAt:     Timestamp(e.CreatedAt),
		ExpiresAt:     Timestamp(e.ExpiresAt),
		RevokedAt:     optNilString(revoked),
	}
}

// isASCIIAlnumByte is u8::is_ascii_alphanumeric.
func isASCIIAlnumByte(b byte) bool {
	return (b >= '0' && b <= '9') || (b >= 'a' && b <= 'z') || (b >= 'A' && b <= 'Z')
}

// vulnerabilityID reports whether id (trimmed) names a finding: 1 to 64
// bytes of ASCII letters, digits, `.`, `_`, `:` and `-`, starting with a
// letter or digit.
func vulnerabilityID(id string) bool {
	if id == "" || len(id) > 64 || !isASCIIAlnumByte(id[0]) {
		return false
	}
	for i := range len(id) {
		b := id[i]
		if !isASCIIAlnumByte(b) && b != '.' && b != '_' && b != ':' && b != '-' {
			return false
		}
	}
	return true
}

// printable checks a trimmed text of 1 to max characters without control
// characters.
func printable(name, value string, maxChars int) error {
	value = trimSpace(value)
	if value == "" || utf8.RuneCountInString(value) > maxChars || strings.IndexFunc(value, unicode.IsControl) >= 0 {
		return kerrors.New(kerrors.Validation, "%s must be 1 to %d printable characters", name, maxChars)
	}
	return nil
}

// checkException is check: the request's limits.
func checkException(body *gen.CreateException) error {
	id := trimSpace(body.Vulnerability)
	if !vulnerabilityID(id) {
		return kerrors.New(kerrors.Validation, "`%s` is not a vulnerability id", id)
	}
	if err := printable("reason", body.Reason, 1024); err != nil {
		return err
	}
	if err := printable("owner", body.Owner, 256); err != nil {
		return err
	}
	const maxDays = scan.MaxExceptionSecs / 86_400
	if body.Days <= 0 || int64(body.Days) > maxDays {
		return kerrors.New(kerrors.Validation, "an exception lasts 1 to %d days", maxDays)
	}
	return nil
}

// accessOrg is a caller and the organization a route acts on for it.
type accessOrg struct {
	access access.Access
	org    ids.OrgID
}

// callerOrg is the organization an org-level route acts on (the caller's
// first, org_of), with the proof of p there.
func (s *Server) callerOrg(ctx context.Context, p perm.Perm, forbidToken bool) (accessOrg, error) {
	a, err := s.access(ctx)
	if err != nil {
		return accessOrg{}, err
	}
	if forbidToken {
		if err := a.ForbidToken(); err != nil {
			return accessOrg{}, err //nolint:wrapcheck // a kerrors already
		}
	}
	org, err := orgOf(a)
	if err != nil {
		return accessOrg{}, err
	}
	if _, err := a.Require(p, authz.OrgChain(org)); err != nil {
		return accessOrg{}, err //nolint:wrapcheck // a kerrors already
	}
	return accessOrg{access: a, org: org}, nil
}

// ListVulnerabilityExceptions is the organization's vulnerability
// exceptions in force (or all).
func (s *Server) ListVulnerabilityExceptions(
	ctx context.Context, params gen.ListVulnerabilityExceptionsParams,
) (gen.ListVulnerabilityExceptionsRes, error) {
	c, err := s.callerOrg(ctx, perm.OrgRead, false)
	if err != nil {
		return nil, err
	}
	t, err := s.deps.Store.Tenant(ctx, c.org)
	if err != nil {
		return nil, err //nolint:wrapcheck // a store error, answered as internal
	}
	defer t.Rollback(ctx) //nolint:errcheck // read only
	now := s.deps.Clock.NowMs()
	found, err := t.Exceptions(ctx, params.All.Or(false))
	if err != nil {
		return nil, err //nolint:wrapcheck // a store error, answered as internal
	}
	out := make(gen.ListVulnerabilityExceptionsOKApplicationJSON, 0, len(found))
	for _, e := range found {
		out = append(out, exceptionDto(e, now))
	}
	return &out, nil
}

// CreateVulnerabilityException lets one finding pass the scan gates for a
// while.
func (s *Server) CreateVulnerabilityException(
	ctx context.Context, req *gen.CreateException,
) (gen.CreateVulnerabilityExceptionRes, error) {
	c, err := s.callerOrg(ctx, perm.OrgAdmin, true)
	if err != nil {
		return nil, err
	}
	if err := checkException(req); err != nil {
		return nil, err
	}
	project := opt.None[ids.ProjectID]()
	if slug, ok := req.Project.Get(); ok {
		p, err := s.findProject(ctx, c.access, slug)
		if err != nil {
			return nil, err
		}
		project = opt.Some(p.project.ID)
	}
	_, actor := c.access.Actor()
	now := s.deps.Clock.NowMs()
	n := store.NewException{
		Project:       project,
		Vulnerability: trimSpace(req.Vulnerability),
		Reason:        trimSpace(req.Reason),
		Owner:         trimSpace(req.Owner),
		CreatedBy:     actor,
		ExpiresAt:     clock.SaturatingAdd(now, int64(req.Days)*86_400_000),
	}
	t, err := s.deps.Store.Tenant(ctx, c.org)
	if err != nil {
		return nil, err //nolint:wrapcheck // a store error, answered as internal
	}
	defer t.Rollback(ctx) //nolint:errcheck // a no-op after the commit
	id, err := t.CreateException(ctx, n)
	if err != nil {
		return nil, err //nolint:wrapcheck // a store error, answered as internal
	}
	audit := requestAudit(c.access, "vulnerability.exception.created", "vulnerability", n.Vulnerability)
	audit.Data = opt.Some[any](map[string]any{"id": id.String(), "owner": n.Owner, "expires_at": n.ExpiresAt})
	if err := t.AppendAudit(ctx, audit); err != nil {
		return nil, err //nolint:wrapcheck // a store error, answered as internal
	}
	all, err := t.Exceptions(ctx, true)
	if err != nil {
		return nil, err //nolint:wrapcheck // a store error, answered as internal
	}
	for _, e := range all {
		if e.ID != id {
			continue
		}
		if err := t.Commit(ctx); err != nil {
			return nil, err //nolint:wrapcheck // a store error, answered as internal
		}
		dto := exceptionDto(e, now)
		return &dto, nil
	}
	return nil, kerrors.New(kerrors.Internal, "the new exception is missing")
}

// RevokeVulnerabilityException revokes an exception for good.
func (s *Server) RevokeVulnerabilityException(
	ctx context.Context, params gen.RevokeVulnerabilityExceptionParams,
) (gen.RevokeVulnerabilityExceptionRes, error) {
	c, err := s.callerOrg(ctx, perm.OrgAdmin, true)
	if err != nil {
		return nil, err
	}
	_, actor := c.access.Actor()
	t, err := s.deps.Store.Tenant(ctx, c.org)
	if err != nil {
		return nil, err //nolint:wrapcheck // a store error, answered as internal
	}
	defer t.Rollback(ctx) //nolint:errcheck // a no-op after the commit
	revoked, err := t.RevokeException(ctx, params.ID, actor)
	if err != nil {
		return nil, err //nolint:wrapcheck // a store error, answered as internal
	}
	if !revoked {
		return nil, kerrors.New(kerrors.NotFound, "exception `%s`", params.ID)
	}
	audit := requestAudit(c.access, "vulnerability.exception.revoked", "vulnerability", params.ID.String())
	if err := t.AppendAudit(ctx, audit); err != nil {
		return nil, err //nolint:wrapcheck // a store error, answered as internal
	}
	if err := t.Commit(ctx); err != nil {
		return nil, err //nolint:wrapcheck // a store error, answered as internal
	}
	return &gen.RevokeVulnerabilityExceptionNoContent{}, nil
}
